### PR TITLE
Fixes #138248, less memory use with beta1=0 python. 

### LINKED ISSUE #138248: Adam optimizer without first moment estimate, for less vram
### 🚀 The feature, motivation and pitch

Many GANs like StyleGan use the Adam optimizer with betas=(0.0, 0.999), this means that the first moment estimate is disabled but it still uses a lot of vram for it, is it possible to add an option to stop this from happening and reduce vram usage?

### Alternatives

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by youssef62**
Fixes #138248. 

Only allocate a `exp_avg` to a tensor of zeros if beta1 > 0; otherwise, set `exp_avg`  to an empty tensor. When doing an optimization step with beta1=0, pass the gradient to `adam` instead of as `exp_avg`. This way, we don't need to change the underlying code to accommodate the above test.

**Tests** : 
* `test_less_mem_beta1_zero_adam`: Make sure memory consumption decreases. 
* `test/test_optim.py:test_correct_beta1` : Make sure adam works correctly with beta1=0. This is done by ensuring it has the same output with `beta1=1e-6` for a few iterations and is different from `beta1=0`. 
* `test_beta1nonzero_then_loadstate_beta1zero_adam`: Ensure we do not break the `load_state_dict`. 
* `test_beta1zero_then_loadstate_beta1nonzero_adam`: Ensure we do not break the `load_state_dict`. 
Finally, added a case with `beta1=0` in `optim_inputs_func_adam`. 

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @mcarilli @ptrblck @leslie-fang-intel

### CODE REVIEW

File: test/test_optim.py [hunk 1]
```diff
@@ -2163,6 +2165,260 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+                case_to_mem_usage["beta1-zero" if beta1_zero else "default"] = sum(
+                    [
+                        item.cuda_memory_usage
+                        if device == "cuda"
+                        else item.cpu_memory_usage
+                        for item in prof.key_averages()
+                    ]
+                )
+
+        self.assertGreater(
```
Comment (janeyx99) [inline test/test_optim.py:2213]: Could we get more precise about memory here? We should be able to figure out the size of exp_avg precisely + test against that.

**Review by janeyx99 — COMMENTED** [review]
Took a brief look + set off CI to see if the tests pass. On the whole everything is the right approach, will review in detail once we get CI signal.

File: test/test_optim.py (hunk 1 above)
Comment (youssef62) [reply test/test_optim.py:2198]: Thanks for your reply. I made the change in my last commit. Let me know if further adjustments are needed.

File: test/test_optim.py [hunk 2]
```diff
@@ -2163,6 +2164,246 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+            for optim_input in optim_info.optim_inputs_func(device=device)
+            if ("betas" in optim_input.kwargs and optim_input.kwargs["betas"][0] == 0.0)
+            or optim_input.desc == "default"
+        ]
+
+        n_iters = 10
+        lr = 0.1
+
+        def run_two_models(beta1_of_default):
+            ouputs = []
```
Comment (janeyx99) [inline test/test_optim.py:2381]: nit: outputs

File: test/test_optim.py [hunk 3]
```diff
@@ -2163,6 +2164,246 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+            for optim_input in optim_inputs:
+                one_model_output = []
+                optim_input.kwargs["lr"] = (
+                    lr  # need lr high enough to see the difference
```
Comment (janeyx99) [inline test/test_optim.py:2385]: could we just use 0.1 here, line 2378 is not really necessary if it's used once!

File: test/test_optim.py [hunk 4]
```diff
@@ -2163,6 +2164,246 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+            if ("betas" in optim_input.kwargs and optim_input.kwargs["betas"][0] == 0.0)
+            or optim_input.desc == "default"
+        ]
+
+        n_iters = 10
+        lr = 0.1
+
+        def run_two_models(beta1_of_default):
+            ouputs = []
+            for optim_input in optim_inputs:
```
Comment (janeyx99) [inline test/test_optim.py:2382]: does this create a model per optim_input that's tested?

File: torch/optim/adam.py [hunk 5]
```diff
@@ -244,7 +250,7 @@ def step(self, closure=None):
             adam(
                 params_with_grad,
                 grads,
-                exp_avgs,
+                exp_avgs if beta1 > 0 else grads,
```
Comment (janeyx99) [inline torch/optim/adam.py:253]: This is a clever approach, but unfortunately the change will have to be more cautious + spotchecked than this, because we _do_ inplace write into the exp_avgs but we never want to inplace modify the grads.

**Review by janeyx99 — COMMENTED** [review]
My most major concern is about the approach --> we can't do the clever solution you added + have to be careful not inplacing into the gradients.

I've also left some review to clean up the testing code a bit.

File: torch/optim/adam.py (hunk 5 above)
Comment (youssef62) [reply torch/optim/adam.py:253]: Thank you for your response. For in-place writes, I believe you're referring to `lerp`. Doesn't it keep exp_avgs constant with a coefficient of 1?

In any case, do you have any suggestions on how to proceed? Whether it's to fix this approach or to explore a new one?

Comment (janeyx99) [reply torch/optim/adam.py:253]: For proceeding, instead of passing grad into exp_avg indiscriminately, I think we should go to every usage of exp_avg and locally assess what to do if beta1 is 0 to ensure that we don't write into it by accident AND to make sure we're not doing unnecessary work. For example, I would guard the following with

if device_beta1 > 0:
    exp_avg.lerp_(grad, 1 - device_beta1)
else:
    exp_avg = grad

https://github.com/pytorch/pytorch/blob/main/torch/optim/adamw.py#L423

Maybe a good first step before this PR is to write a test making sure grad is never inplaced into...let me do that right now actually!

Comment (janeyx99) [reply torch/optim/adam.py:253]: https://github.com/pytorch/pytorch/pull/143612 done here, i'll try merging this soon

File: test/test_optim.py [hunk 6]
```diff
@@ -2199,6 +2200,202 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+            )
+            if beta1_zero or optim_input.desc == "default":
+                optim = optim_info.optim_cls(model.parameters(), **optim_input.kwargs)
+                optim.zero_grad() 
+                output = model(inpt)
+                loss = output.sum()
+                loss.backward()
+                optim.step()
+
+                if not beta1_zero :
```
Comment (janeyx99) [inline test/test_optim.py:2239]: isn't this not going to be true since it's under the conditional in 2231?

Either way, this test can be simplified into the following:
```
    @optims(
        [optim for optim in optim_db if optim.optim_cls.__name__ == "Adam"],
        dtypes=[torch.float32],
    )
    def test_less_mem_beta1_zero_adam(self, device, dtype, optim_info):
        # test that beta1=0.0 uses less memory than the default
        beta_1_optim_inputs = [o for o in optim_info.optim_inputs_func(device=device) if "betas" in optim_input.kwargs and optim_input.kwargs["betas"][0] > 0.0]

        num_params = 4
        size_of_param_in_bytes = 32 * 16 * sizeof(dtype)  # hardcoded, also idk if sizeof is the right api)
        for optim_input in beta_1_optim_inputs:
            beta1_values = (optim_input.kwargs["betas"][0], 0.0)
            total_sizes = []    # will end up as [big_state_dict_size, no_exp_avg_sd_size]
            for beta1 in beta1_values:
                optim_input.kwargs["betas"] = (beta1, optim_input.kwargs["betas"][1])
                params = [torch.rand(32, 16, device=device, dtype=dtype) for _ in range(num_params)]
                for p in params:
                    p.grad = torch.rand_like(p)

                optim = optim_info.optim_cls(params, **optim_input.kwargs)
                for _ in range(3):
                    optim.step()
                    optim.zero_grad()
                    for p in params:
                        p.grad = torch.rand_like(p)

                total_sizes.append(get_obj_size(optim.state_dict()))

            # size of exp_avg should match the size of params, as all the params here requires grad
            size_of_exp_avgs = num_params * size_of_param_in_bytes
            self.assertEqual(total_sizes[0], total_sizes[1] + size_of_exp_avgs)

File: test/test_optim.py (hunk 4 above)
Comment (janeyx99) [reply test/test_optim.py:2375]: If there are only 2 inputs, it may be better to not use the optim_info.optim_inputs_func(device=device) call but to just explicitly create two optimizers for clarity.

like optim_with_beta1 = ...
optim_no_beta1 = 

optim_inputs = [optim_with_beta1, optim_no_beta1]

rest of test is the same

File: test/test_optim.py [hunk 7]
```diff
@@ -2199,6 +2200,202 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+        
+
+    @optims(
+        [optim for optim in optim_db if optim.optim_cls.__name__ == "Adam"],
+        dtypes=[torch.float32],
+    )
+    def test_loadstate_with_diff_beta1_adam(
+        self, device, dtype, optim_info
+    ):
+        # the test takes form of two tests.
```
Comment (janeyx99) [inline test/test_optim.py:2263]: What exactly is this test testing? It looks like it tests that loading a beta1=0 optim into a default optim will compute similar results, which doesn't have to be true, right?

Could you see if you could simplify the test more to make it more readable?

One way to do so:
1. create two optimizers, one with beta1 is zero, the other nonzero
2. run both for a few steps
3. switch their state!
4. run both again for a few steps

There's no need to assert equality, as they're not expected to be equal to each other, just that they can be loaded into each other!

**Comment by janeyx99** [comment]
The approach looks good to me if the tests pass. I do have comments on the tests as I believe you can simplify them a bit!

File: aten/src/ATen/native/cpu/FusedAdamKernel.cpp [hunk 8]
```diff
@@ -238,13 +238,19 @@ std::enable_if_t<
...
+    if (exp_avg_grad_coefficient < 1){
+      auto is_lerp_weight_small = std::abs(scalar_t(exp_avg_grad_coefficient)) < scalar_t(0.5);
+      if (is_lerp_weight_small) {
+        exp_avg_ptr[d] = exp_avg_ptr[d] + scalar_t(exp_avg_grad_coefficient) * (grad_val - exp_avg_ptr[d]);
+      } else {
+        exp_avg_ptr[d] = grad_val - (grad_val - exp_avg_ptr[d]) * (scalar_t(1) - scalar_t(exp_avg_grad_coefficient));
+      }
+    }
+    else{
+      exp_avg_ptr[d] = grad_ptr[d];
```
Comment (youssef62) [inline aten/src/ATen/native/cpu/FusedAdamKernel.cpp:251]: `else exp_avg = grad` because `adam_math` is used by AdamW so I added it to keep compatibility.

**Comment by youssef62** [comment]
I missed that AdamW also used `adam_math` of `aten/src/ATen/native/cpu/FusedAdamKernel.cpp`. I adapted it accordingly, could you set off CI again? 

I also simplified tests as suggested, only one left `test_loadstate_with_diff_beta1_adam` for which I asked for clarifications above.

File: aten/src/ATen/native/cpu/FusedAdamKernel.cpp [hunk 9]
```diff
@@ -238,7 +238,7 @@ std::enable_if_t<
         param_ptr[d] = param_ptr[d] * scalar_t(1 - lr * weight_decay);
       }
     }
-    // exp_avg.lerp_(grad, 1 - beta1) if beta1 > 0 else exp_avg = grad 
+    // exp_avg.lerp_(grad, 1 - beta1) if beta1 > 0 else grad
```
Comment (janeyx99) [inline aten/src/ATen/native/cpu/FusedAdamKernel.cpp:241]: why the comment change? what you had before was clearer! 
```suggestion
    // exp_avg.lerp_(grad, 1 - beta1) if beta1 > 0 else exp_avg = grad  
```

File: aten/src/ATen/native/cpu/FusedAdamKernel.cpp (hunk 8 above)
Comment (janeyx99) [reply aten/src/ATen/native/cpu/FusedAdamKernel.cpp:251]: Is this a no-op tho? If this effectively does 

```suggestion
      grad_ptr[d] = grad_ptr[d];
```

Then I would just leave a comment and get rid of the else block as it would be doing unnecessary work

File: test/test_optim.py [hunk 10]
```diff
@@ -1334,6 +1334,7 @@ def closure():
     @optims(optim_db, dtypes=[torch.float32])
     def test_grads_are_never_inplaced_into(self, device, dtype, optim_info):
         optim_cls = optim_info.optim_cls
+
```
Comment (janeyx99) [inline test/test_optim.py:1337]:
```suggestion
```

why the white line changes? they make it ever so slightly harder to review 😛

File: test/test_optim.py [hunk 11]
```diff
@@ -2205,69 +2206,66 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+        beta_1_optim_inputs = [
+            o
+            for o in optim_info.optim_inputs_func(device=device)
+            if "betas" in o.kwargs and o.kwargs["betas"][0] > 0.0
+        ]
+
+        num_params = 4
+        print(f"num_params: {num_params}")
+        print(f"sizeof: {32 * 16 * dtype.__sizeof__()}")
+        size_of_param_in_bytes = (
```
Comment (janeyx99) [inline test/test_optim.py:2228]: ah, you can also compute this later when you have the params, so like

num_params * params[0].nelement() * params[0].element_size()

File: test/test_optim.py [hunk 12]
```diff
@@ -2361,38 +2347,38 @@ def step(model, optim):
...
-            or optim_input.desc == "default"
-        ]
 
-        n_iters = 10
-        self.assertEqual(len(optim_inputs), 2) # we need two optimizers to compare
-        def run_two_models(beta1_of_default):
+        def run_two_models(non_zero_beta1):
             outputs = []
+            # lr of .1 needed to see difference.
+            optim_inputs = [
```
Comment (janeyx99) [inline test/test_optim.py:2354]: rename to optim_kwargs for clarity

File: test/test_optim.py (hunk 7 above)
Comment (janeyx99) [reply test/test_optim.py]: If I understand the test currently written, then it is doing 

first N steps:
params A get updates with a nonzero beta
params B get updates with zero beta

then second N steps:
params A get updates with zero beta (since it got swapped)
params B get updates with nonzero beta

which I don't think is equivalent. I DO think the following is equivalent:

first N steps:
params A get updates with a nonzero beta

then we load a state dict into existing optim A and a new optim B both now with beta

then we run for more steps and check that those are the same.

File: aten/src/ATen/native/cpu/FusedAdamKernel.cpp (hunk 8 above)
Comment (youssef62) [reply aten/src/ATen/native/cpu/FusedAdamKernel.cpp:251]: It maintains compatibility with AdamW when beta1=0.

Comment (janeyx99) [reply aten/src/ATen/native/cpu/FusedAdamKernel.cpp:251]: Ah, would you be okay with changing this PR to update both then? We recently landed a PR that combined AdamW into Adam so it should not be too much more code: https://github.com/pytorch/pytorch/pull/143710

File: test/test_optim.py (hunk 7 above)
Comment (youssef62) [reply test/test_optim.py]: > first N steps:
params A get updates with a nonzero beta
params B get updates with zero beta
then second N steps:
params A get updates with zero beta (since it got swapped)
params B get updates with nonzero beta

I do not do any swapping. I do something more like what you explained after. More exactly : 

First N steps : 
params A get updates with a nonzero beta
params B : nothing 

Load A with a state that has beta = 0. 

Second N steps: 
params A get updates with zero beta.  
params B get updates with zero beta (its initial beta). 

At the end I expect A and B to behave the same.

Comment (janeyx99) [reply test/test_optim.py]: ah but the params in A and B need to be matching between the first and second N steps, right?

File: torch/optim/adam.py (hunk 5 above)
Comment (youssef62) [reply torch/optim/adam.py:256]: `exp_avgs if beta1 > 0 else grads` provoques ` Dynamic control flow is not supported at the moment. Please use torch.cond to explicitly capture the control flow. For more information about this error` for the compiled version. The two possibility I can think of is either replace it with `torch.cond(beta1 > 0, lambda: exp_avgs, lambda: grads)`. Do you think this is a good fix ? As opposed to adding an if-else before the `adam` call.

File: test/test_optim.py (hunk 10 above)
Comment (youssef62) [reply test/test_optim.py:1348]: Sorry, will be more careful next time.

**Comment by janeyx99** [comment]
ci failures are real, it looks like we need to maintain tensorness of betas if betas are passes as tensors

File: torch/optim/adam.py (hunk 5 above)
Comment (janeyx99) [reply torch/optim/adam.py:256]: Yea this is okay

File: torch/optim/adam.py [hunk 13]
```diff
@@ -174,9 +175,14 @@ def _init_group(
...
                     )
                     # Exponential moving average of gradient values
-                    state["exp_avg"] = torch.zeros_like(
-                        p, memory_format=torch.preserve_format
+                    # case beta1 == 0, we don't need exp_avg
+
+                    state["exp_avg"] = (
+                        torch.zeros_like(p, memory_format=torch.preserve_format)
+                        if beta1 > 0
+                        else torch.zeros(0)
```
Comment (janeyx99) [inline torch/optim/adam.py:183]:
```suggestion
                        else None
```
Wait why do we have this entry at all if beta1 is 0? We should make exp_avg optionally exist in the implementations.

File: test/test_optim.py [hunk 14]
```diff
@@ -2216,6 +2218,188 @@ def test_non_empty_state(self, device, dtype, optim_info):
...
+            self.assertEqual(total_sizes[0], total_sizes[1] + size_of_exp_avgs)
+
+    @optims(
+        [optim for optim in optim_db if optim.optim_cls.__name__ in ["Adam", "AdamW"]],
+        dtypes=[torch.float32],
+    )
+    def test_loadstate_with_diff_beta1_adam(self, device, dtype, optim_info):
+        # the test takes form of two tests.
+        # 1st test.
+        # model_beta1_zero: 1.train for n_iters, 2.load state, 3.train for n_iters
```
Comment (janeyx99) [inline test/test_optim.py:2281]:
```suggestion
        # model_beta1_zero: 1.train for n_iters, 2.load state of default, 3.train for n_iters
```

**Comment by youssef62** [comment]
Hello, I am back working on this. Sorry for making the pr longer than needed. Dynamo is giving me a hard time, as I have many dynamic if conditions. Is there a way to have CI go off automatically when I push ? It would help me iterate faster.

**Comment by janeyx99** [comment]
@youssef62 unfortunately not (as for security reasons the repo only allows CI for folks with write access), but just ping me every time and i will try to be timely about it!

**Comment by youssef62** [comment]
@janeyx99 Could you run the CI ?

**Comment by youssef62** [comment]
@janeyx99  I messed up a lot with git in this PR, do you mind if I create a new PR with the same code ?

**Comment by janeyx99** [comment]
@youssef62 yes go for it!
